File: OS_CP_Phase_2/src/mos/OS.java
package mos;

import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.FileInputStream;
import java.io.FileReader;
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.Writer;
import java.io.OutputStreamWriter;
import java.io.FileOutputStream;
import java.util.*;

public class OS {

  // Essential variables
  private char[][] M = new char[300][4];
  private char[][] PT = new char[300][4];
  private char[] IR = new char[4];
  private char[] R = new char[4];
  private int SI = 3;
  private int TI = 0;
  private int PI;
  private int PTR;
  private String VA;
  private int RA;
  private int TTC;
  private int LLC;
  private int TTL;
  private int TLL;
  private int JID;
  private int EM = 0;
  private int IC;
  private boolean C = false;
  private char[] inputBuffer = new char[40];
  // private int data_index = 0;
  // private String time, id, lines_printed;
  private String inputFile;
  private String outputFile;

  // Semi-Essential variables
  private int memory_used;
  private FileReader input;
  private BufferedReader buff;
  private FileWriter output;
  Map<Integer, String> map = new HashMap<>();
  private int BA = 0;

  // Constructor
  public OS(String file, String output) {
    this.inputFile = file;
    this.outputFile = output;
    this.SI = 0;

    try {
      this.input = new FileReader(inputFile);
      this.buff = new BufferedReader(input);
      this.output = new FileWriter(output);
    } catch (IOException e) {
      e.printStackTrace();
    }

    map.put(++EM, "OUT OF DATA");
    map.put(++EM, "LINE LIMIT EXCEEDED");
    map.put(++EM, "TIME LIMIT EXCEEDED");
    map.put(++EM, "OPERATION CODE");
    map.put(++EM, "OPERAND");
    map.put(++EM, "INVALID PAGE FAULT");
    map.put(++EM, "TIME LIMIT EXCEEDED AND OPERATION CODE ERROR");
    map.put(++EM, "TIME LIMIT EXCEEDED AND OPERAND ERROR");
    EM = 0;
  }

  public void print_memory() {
    for (int i = 0; i < 300; i++) {
      System.out.println("memory[" + i + "] " + new String(M[i]));
    }
  }

  public void OUTPUT() throws IOException {
    FileWriter fr = new FileWriter(outputFile);
    for (int i = 0; i < 300; i++) {
      for (int j = 0; j < 4; j++) {
        try {
          fr.write(M[i][j]);
        } catch (IOException e) {
          System.out.println("Error in output: " + e);
        }
      }
    }
    fr.close();
  }

  public void EXECUTE_USER_PROGRAM() throws Exception {
    while (2 < 3) {
      if (IC == 300) {
        break;
      }

      VA = String.valueOf(IC);
      RA = ADDRESSMAP(VA);
      IR[0] = M[RA][0];
      IR[1] = M[RA][1];
      IR[2] = M[RA][2];
      IR[3] = M[RA][3];
      IC++;
      StringBuilder v = new StringBuilder();
      v.append(IR[2]).append(IR[3]);
      String vb = v.toString();
      VA = String.valueOf(vb);
      RA = ADDRESSMAP(VA);

      if (IR[0] == 'L' && IR[1] == 'R') {
        String line = new String(IR);
        // System.out.println(line);
        // System.out.println(line.substring(2));
        int num = Integer.parseInt(line.substring(2));
        R[0] = M[num][0];
        R[1] = M[num][1];
        R[2] = M[num][2];
        R[3] = M[num][3];
      } else if (IR[0] == 'S' && IR[1] == 'R') {
        String line = new String(IR);
        int num = Integer.parseInt(line.substring(2));
        M[num][0] = R[0];
        M[num][1] = R[1];
        M[num][2] = R[2];
        M[num][3] = R[3];
      } else if (IR[0] == 'C' && IR[1] == 'R') {
        String line = new String(IR);
        int num = Integer.parseInt(line.substring(2));
        if (M[num][0] == R[0] && M[num][1] == R[1] && M[num][2] == R[2] && M[num][3] == R[3]) {
          C = true;
        } else {
          C = false;

        }
      } else if (IR[0] == 'B' && IR[1] == 'T') {
        if (C == true) {
          String line = new String(IR);
          int num = Integer.parseInt(line.substring(2));
          IC = num;
          C = false;
        }
      } else if (IR[0] == 'G' && IR[1] == 'D') {
        SI = 1;
        MOS(inputBuffer);
      } else if (IR[0] == 'P' && IR[1] == 'D') {
        SI = 2;
        MOS(inputBuffer);
      } else if (IR[0] == 'H') {
        SI = 3;
        MOS(inputBuffer);
      } else {
        PI = 1;
        MOS(inputBuffer);
      }
      SIMULATION();
    }
  }

  public static boolean isNumeric(String str) {
    try {
      Double.parseDouble(str);
      return true;
    } catch (NumberFormatException e) {
      return false;
    }
  }

  public int ADDRESSMAP(String va) throws Exception {
    int ra = 0;
    if (isNumeric(va)) {
      char[] frameno = new char[2];
      frameno[0] = (char) (Character.getNumericValue(M[PTR][2]));
      frameno[1] = (char) (Character.getNumericValue(M[PTR][3]));
      String r = new String(frameno);

      if (frameno[0] == '*' && frameno[1] == '*') {
        PI = 3;
        MOS(inputBuffer);
      } else {
        ra = Integer.parseInt(r) * 10 + (Integer.parseInt(va) / 10);
      }
    } else {
      PI = 2;
      MOS(inputBuffer);
    }

    return ra;
  }

  public void SIMULATION() throws Exception {
    TTC++;
    if (TTC > TTL) {
      TI = 2;
      MOS(inputBuffer);
    }
  }

  public void READ(char[] inputBuffer) throws Exception {

    if (inputBuffer[0] == '$' && inputBuffer[1] == 'E' && inputBuffer[2] == 'N' && inputBuffer[3] == 'D') {
      TERMINATE(1);
    }

    IR[3] = '0';
    String line = new String(IR);

    int num = Integer.parseInt(line.substring(2)); // operand

    try {
      line = buff.readLine();
    } catch (IOException e) {
      e.printStackTrace();
    }

    inputBuffer = line.toCharArray();
    for (int i = 0; i < line.length();) {
      M[num][i % 4] = inputBuffer[i];
      i++;
      if (i % 4 == 0) {
        num++;
      }
    }
  }

  public void WRITE() throws Exception {
    LLC++;
    if (LLC > TLL) {
      TERMINATE(2);
    }

    IR[3] = '0';
    String t, total = "";
    String line = new String(IR);
    int num = Integer.parseInt(line.substring(2));
    for (int i = 0; i < 10; i++) {
      t = new String(M[num + i]);
      total = total.concat(t);
    }
    System.out.println(total + "In write");
    try {
      Writer output = new BufferedWriter(new OutputStreamWriter(
          new FileOutputStream(outputFile, true), "UTF-8"));
      output.append(total);
      output.append("\r\n");
      output.close();
    } catch (IOException e) {
      e.printStackTrace();
    }
    EXECUTE_USER_PROGRAM();
  }

  public void TERMINATE(int error) throws Exception {
    Writer output;
    output = new BufferedWriter(new FileWriter(outputFile, true));
    output.append('\n');
    output.append('\n');
    System.out.println("Program Executed");

    if (error > 0) {
      System.out.println("JOB ID : " + JID);
      System.out.println(map.get(error));
      System.out.println("IC : " + IC);
      String ir = new String(IR);
      System.out.println("IR : " + Integer.parseInt(ir));
      System.out.println("TTC : " + TTC);
      System.out.println("LLC : " + LLC);
    }
    output.close();
    LOAD();
  }

  public void LOAD() throws Exception {

    SI = 0;
    String strLine;

    try {
      // This loop fills the 40 byte inputBuffer for buffered execution
      while ((strLine = buff.readLine()) != null) {
        inputBuffer = strLine.toCharArray(); // Started filling inputBuffer

        if (inputBuffer[0] == '$' && inputBuffer[1] == 'A' && inputBuffer[2] == 'M' && inputBuffer[3] == 'J') {
          init();
          StringBuilder sb = new StringBuilder();
          sb.append(inputBuffer[4]).append(inputBuffer[5]).append(inputBuffer[6]).append(inputBuffer[7]);
          int jid = Integer.parseInt(sb.toString());
          sb.setLength(0);
          sb.append(inputBuffer[8]).append(inputBuffer[9]).append(inputBuffer[10]).append(inputBuffer[11]);
          int ttl = Integer.parseInt(sb.toString());
          sb.setLength(0);
          sb.append(inputBuffer[12]).append(inputBuffer[13]).append(inputBuffer[14]).append(inputBuffer[15]);
          int tll = Integer.parseInt(sb.toString());
          sb.setLength(0);
          PCB pcb = new PCB(jid, ttl, tll);
          JID = pcb.JID;
          TTL = pcb.TLL;
          TLL = pcb.TLL;
          PTR = ALLOCATE();
          int row = PTR;
          int col = 0;
          while (row < PTR + 10) {
            PT[row][col] = '0';
            row++;
          }
          while (row < PTR + 10) {
            col = 1;
            PT[row][col] = '*';
            row++;
          }
          while (row < PTR + 10) {
            col = 2;
            PT[row][col] = '*';
            row++;
          }
          continue;
        }

        else if (inputBuffer[0] == 'G' && inputBuffer[1] == 'D' || inputBuffer[0] == 'P' && inputBuffer[1] == 'D'
            || inputBuffer[0] == 'L' && inputBuffer[1] == 'R' || inputBuffer[0] == 'S' && inputBuffer[1] == 'R'
            || inputBuffer[0] == 'B' && inputBuffer[1] == 'T' || inputBuffer[0] == 'C' && inputBuffer[1] == 'R'
            || inputBuffer[0] == 'H') {
          BA = ALLOCATE();
          int digit = BA / 10;
          int row = PTR;
          String cc = Integer.toString(digit);
          if (cc.length() == 1) {
            PT[row][2] = cc.charAt(0);
            PT[row][1] = '0';
          } else {
            PT[row][1] = cc.charAt(0);
            PT[row][2] = cc.charAt(1);
          }

          String tem = new String(inputBuffer);
          System.out.println(tem);

          for (int k = BA; k < BA + 10;) {
            for (int j = 0; j < strLine.length();) {
              M[k][j % 4] = tem.charAt(j);
              j++;
              if (j % 4 == 0)
                k++;
            }
          }
        }

        else if (inputBuffer[0] == '$' && inputBuffer[1] == 'D' && inputBuffer[2] == 'T' && inputBuffer[3] == 'A') {
          PTR++;
          EXECUTE_USER_PROGRAM();
          continue;
        } else if (inputBuffer[0] == '$' && inputBuffer[1] == 'E' && inputBuffer[2] == 'N' && inputBuffer[3] == 'D') {
          System.out.println("\n\n\n");
          SI = 3;
          MOS(inputBuffer);
          continue;
        }
        if (IC == 300) { // abort;
          System.out.println("Abort due to exceed M usage");
        }
      }
    } catch (Exception e) {
      e.printStackTrace();
    }
  }

  public int ALLOCATE() {
    int size = 30;

    ArrayList<Integer> list = new ArrayList<Integer>(size);
    for (int i = 0; i < size; i++) {
      list.add(i);
    }

    int temp = 0;

    Random rand = new Random();
    while (list.size() > 0) {
      int index = rand.nextInt(list.size());
      temp = list.remove(index);
    }

    return temp * 10;
  }

  public void START_EXECUTION() throws Exception {
    IC = 00;
    EXECUTE_USER_PROGRAM();
  }

  public void init() {
    BA = 0;
    M = new char[300][4];
    C = false;
    IC = 0;
  }

  public void MOS(char[] inputBuffer) throws Exception {
    int si = this.SI;
    int ti = this.TI;
    int pi = this.PI;

    // Ti and Si section

    if (ti == 0 && si == 1) {
      READ(inputBuffer);
    }
    if (ti == 0 && si == 2) {
      WRITE();
    }
    if (ti == 0 && si == 3) {
      TERMINATE(0);
    }
    if (ti == 2 && si == 1) {
      TERMINATE(1);
    }
    if (ti == 2 && si == 2) {
      TERMINATE(3);
    }
    if (ti == 2 && si == 3) {
      TERMINATE(0);
    }

    // Ti and Pi section

    if (ti == 0 && pi == 1) {
      TERMINATE(4);
    }

    if (ti == 0 && pi == 2) {
      TERMINATE(5);
    }

    if (ti == 0 && pi == 3) {
      // if valid handle using paging
      TERMINATE(6);
    }

    if (ti == 2 && pi == 1) {
      TERMINATE(7);
    }

    if (ti == 2 && pi == 2) {
      TERMINATE(8);
    }

    if (ti == 2 && pi == 3) {
      TERMINATE(3);
    }

  }
}

File: OS_CP_Phase_2/OS.py
import random
import os
os.chdir(r'D:\ISHAN\TY_AI-A_Sem 5\OS\CP\OS-CP\OS_CP_Phase_2')

m = 0
IR = [0 for i in range(4)]
IC = [0 for i in range(2)]
R = [0 for i in range(4)]
C = False
SI = 0
PI = 0
TI = 0
PTR = [0 for i in range(4)]

used_frames = set()

M = [['\0' for i in range(4)] for j in range(300)]
opfile = open('output.txt', 'w')
input_buffer = []  # size is 40 bytes
data_index = 0
pd_error = 0
gd_error = 0


class PCB:
    def __init__(self, job_id, ttl, tll, ttc, llc):
        self.job_id = job_id
        self.TTL = ttl   # total time limit
        self.TTC = ttc   # total time counter
        self.TLL = tll   # total line limit
        self.LLC = llc   # line limit counter

    def incrementLLC(self):
        self.LLC = self.LLC + 1

    def incrementTTC(self):
        self.TTC = self.TTC + 1


def print_memory():
    for i in range(300):
        for j in range(4):
            print(M[i][j], end=" ", file=opfile)
        print("\n", file=opfile)


def READ(address):
    global data_index, gd_error
    i, j = 0, 0
    line = input_buffer[data_index]
    if (line[0] == "$" and line[1] == "E"):  # checking for out of data error
        gd_error = -1
        return
    address = (address // 10) * 10
    while (line[i] != '\n'):          # adding data to the M at received address
        M[address][j] = line[i]
        i += 1
        j += 1
        if (j == 4):
            j = 0
            address += 1
    data_index += 1


def WRITE(address):
    global pd_error
    if (pcb.LLC >= pcb.TLL):  # checking for line limit exceeded
        pd_error = -1
        return
    address = (address // 10) * 10

    for i in range(address, address + 10):   # printing entire block to file
        for j in range(4):
            if (M[i][j] == '\0'):
                break
            print(M[i][j], end="", file=opfile)
    print('\n', file=opfile)
    pcb.incrementLLC()


def TERMINATE(EM):
    JID = pcb.job_id
    print("JOB ID : ", JID, file=opfile)
    if (EM == 0):
        print("\tProgram Executed Sucessfully", file=opfile)
    elif (EM == 1):
        print("\tOut of Data Error", file=opfile)
    elif (EM == 2):
        print("\tLine Limit Exceeded", file=opfile)
    elif (EM == 3):
        print("\tTime Limit Exceeded", file=opfile)
    elif (EM == 4):
        print("\tOperation Code Error", file=opfile)
    elif (EM == 5):
        print("\tOperand Error", file=opfile)
    elif (EM == 6):
        print("\tInvalid Page Fault", file=opfile)
    elif (EM == 7):
        print("\tTLE with opcode error", file=opfile)
    elif (EM == 8):
        print("\tTLE with operand error", file=opfile)

    print("IC     : ", IC, file=opfile)
    print("IR     : ", IR, file=opfile)
    print("TTC    : ", pcb.TTC, file=opfile)
    print("LLC    : ", pcb.LLC, file=opfile)
    print("TTL    : ", pcb.TLL, file=opfile)
    print("TTL    : ", pcb.TTL, file=opfile)
    print("\n", file=opfile)


def ALLOCATE():
    frame = (random.randint(0, 29))
    return frame


def LOAD():
    with open("input.txt", "r") as file:
        global input_buffer, M

        input_buffer = file.readlines()
        counter = -1
        index = 0
        while (index < len(input_buffer)):
            global data_index, pcb, id, time, lines_printed, PTR, used_frames, M, C, IR, IC, R, SI, PI, TI, pd_error, gd_error
            line = input_buffer[index]
            if (line[0].startswith('$')):
                if (line[1:4] == 'AMJ'):  # control card
                    id = line[4:8]
                    time = line[8:12]
                    lines_printed = line[12:16]
                    counter = 0
                    pcb = PCB(id, int(time), int(lines_printed), 0,
                              0)  # initialize PCB
                    # initialise frame for page table
                    frame_num = ALLOCATE()
                    used_frames.add(frame_num)   # add frame to used frames set
                    frame_num *= 10
                    # initialise page table register
                    PTR[1] = frame_num // 100
                    frame_num = frame_num % 100
                    PTR[2] = frame_num // 10
                    PTR[3] = frame_num % 10

                elif (line[1:4] == 'DTA'):
                    counter = 1
                    data_index = index + 1  # index of where data address begins
                    START_EXECUTION()
                    index = data_index - 1

                elif (line[1:4] == 'END'):
                    counter = -1
                    print("new prog")
                    # resetting all globals
                    C = False
                    M = [['\0' for i in range(4)] for j in range(300)]

                    IR = [0 for i in range(4)]
                    IC = [0 for i in range(2)]
                    R = [0 for i in range(4)]
                    C = False
                    SI = 0  # need to be looked
                    pd_error = 0
                    gd_error = 0
                    PI = 0
                    TI = 0
                    PTR = [0 for i in range(4)]
                    used_frames.clear()
                    counter = -1
                else:
                    print('error in input')
                    exit(-1)

            else:
                if (counter == 0):  # for reading instructions
                    # initialising frame for program
                    frame_num_prog = ALLOCATE()
                    while frame_num_prog in used_frames:  # finding unique frame
                        frame_num_prog = ALLOCATE()
                    used_frames.add(frame_num_prog)

                    pt_num = int(PTR[1]) * 100 + int(PTR[2]) * 10 + int(
                        PTR[3])  # updating page table entry

                    M[pt_num][2] = frame_num_prog // 10
                    M[pt_num][3] = frame_num_prog % 10
                    M[pt_num][0] = 0
                    M[pt_num][1] = 0

                    i = 0
                    # print(int(time))

                    frame_num_p = frame_num_prog * 10  # address where program is stored

                    while (i < len(line)):

                        if (line[i] == 'H'):

                            M[frame_num_p][0] = line[i]
                            i += 1  # since H has no operands associated with it
                            frame_num_p += 1
                        else:
                            M[frame_num_p][0:4] = line[i:i + 4]

                            i += 4
                            frame_num_p += 1
                        if (frame_num_p % 10 == 0):  # if frame size exceeded- assign new frame
                            frame_num_prog = ALLOCATE()
                            while frame_num_prog in used_frames:
                                frame_num_prog = ALLOCATE()
                            used_frames.add(frame_num_prog)
                            frame_num_p = frame_num_prog * 10
                            pt_num += 1  # increment page table index
                            # update page table entry
                            M[pt_num][2] = frame_num_prog // 10
                            M[pt_num][3] = frame_num_prog % 10
                            M[pt_num][0] = 0
                            M[pt_num][1] = 0
            index += 1


def START_EXECUTION():
    IC[0] = 0
    IC[1] = 0
    EXECUTE_USER_PROGRAM()


def MOS(valid=False):
    global SI, TI, PI, gd_error, pd_error

    if (gd_error == -1):
        TERMINATE(1)

    if (pd_error == -1):
        TERMINATE(2)

    if (TI == 0):
        if (PI == 1):
            TERMINATE(4)  # Operation Code Error
        elif (PI == 2):
            TERMINATE(5)  # Operand Error
        elif (PI == 3):  # page fault
            if (valid):  # valid argument passed to master mode function
                valid_page_fault()
            else:
                TERMINATE(6)  # invalid page fault
        else:
            if (SI == 1):  # read function GD
                READ(ADDRESSMAP(int(IR[2]) * 10 + int(IR[3])))
            elif (SI == 2):  # write function PD
                WRITE(ADDRESSMAP(int(IR[2]) * 10 + int(IR[3])))
            elif (SI == 3):  # TERMINATE successfully
                TERMINATE(0)

    elif (TI == 2):
        if (PI == 1):
            TERMINATE(7)  # TLE with opcode error
        elif (PI == 2):
            TERMINATE(8)  # TLE with operand error
        elif (PI == 3):
            TERMINATE(3)  # Time Limit Exceeded

    else:
        if (SI == 1):
            TERMINATE(3)
        elif (SI == 2):
            WRITE(ADDRESSMAP(int(IR[2]) * 10 + int(IR[3])))
            TERMINATE(3)
        elif (SI == 3):
            TERMINATE(0)


def ADDRESSMAP(VA):

    global PTR, M
    # first we get page table entry
    pte = (int(PTR[1]) * 100 + int(PTR[2]) * 10 + int(PTR[3])) + VA // 10

    if M[pte][0] == '\0':  # checking if anything is present in page table entry
        return -1  # if not invoke page fault

    # calculate M frame entry from pte
    addr = int(M[pte][2]) * 10 + int(M[pte][3])
    RA = addr * 10 + VA % 10  # calculate real address
    return RA


def valid_page_fault():
    global used_frames, M, PTR
    frame_num_data = ALLOCATE()  # initialise a new frame for data
    while frame_num_data in used_frames:
        frame_num_data = ALLOCATE()
    used_frames.add(frame_num_data)
    # find page table index which is empty
    c_ptr = (int(PTR[1]) * 100 + int(PTR[2]) * 10 + int(PTR[3]))
    while (M[c_ptr][0] != '\0'):
        c_ptr += 1
    # add page table entry
    M[c_ptr][2] = frame_num_data // 10
    M[c_ptr][3] = frame_num_data % 10
    M[c_ptr][0] = 0
    M[c_ptr][1] = 0

    print(M[c_ptr])


def SIMULATION():
    global SI, TI
    pcb.incrementTTC()
    if (pcb.TTC > pcb.TTL):
        SI = 1
        TI = 2
        MOS()


def EXECUTE_USER_PROGRAM():
    while (1):
        global IC, IR, R, C, T, SI, TI, PI
        SI = 0
        PI = 0
        TI = 0

        # converting virtual address to real addresss
        inst_count = ADDRESSMAP(10 * IC[0] + IC[1])
        if (inst_count == -1):  # master mode - operand error
            PI = 2
            MOS()
            break

        IC[1] += 1  # incrementing IC
        if IC[1] == 10:
            IC[0] += 1
            IC[1] = 0

        IR = M[inst_count]
        print("IR", IR)

        inst = "" + IR[0] + IR[1]

        if (inst[0] != "H"):

            if ((IR[2].isnumeric() and IR[3].isnumeric()) == False):  # checking for operand error
                PI = 2
                MOS()
                break
            real_address = ADDRESSMAP(int(IR[2]) * 10 + int(IR[3]))

        if inst == "LR":
            if (real_address == -1):  # invalid page fault
                PI = 3
                MOS()
                break
            R = M[real_address]

        elif inst == "SR":
            if (real_address == -1):  # valid page fault
                PI = 3
                # decrementing IC
                if IC[1] != 0:
                    IC[1] -= 1
                else:
                    IC[1] = 9
                    IC[0] -= 1
                MOS(valid=True)
                pcb.TTC -= 1
                # pcb.TTC -= 1
                continue
            else:
                M[real_address] = R

        elif inst == "CR":

            if (real_address == -1):  # invalid page fault
                PI = 3
                MOS()
                break
            if (M[real_address] == R):
                C = True
            else:
                C = False

        elif inst == "BT":
            if (C):
                IC[0] = int(IR[2])
                IC[1] = int(IR[3])

        elif inst == "GD":
            print(inst_count, real_address)
            if (real_address == -1):  # valid page fault
                PI = 3
                SI = 0
                print('valid page fault')
                MOS(valid=True)
                if IC[1] != 0:
                    IC[1] -= 1
                else:
                    IC[1] = 9
                    IC[0] -= 1
                PI = 0
                pcb.TTC -= 1
                continue
            SI = 1
            MOS()
            if (gd_error == -1):
                MOS()
                break
        elif inst == "PD":
            if (real_address == -1):  # invalid page fault
                PI = 3
                MOS()
                break
            else:
                SI = 2
                MOS()
                if (pd_error == -1):
                    MOS()
                    break

        elif inst == "H\0":
            SI = 3
            MOS()
            break

        else:
            PI = 1
            MOS()
            break
        SIMULATION()
    pcb.incrementTTC


if __name__ == "__main__":
    LOAD()
    opfile.close()
